apps/emqx_management/src/emqx_mgmt_data_backup.erl

```diff
@@ -0,0 +1,674 @@
+        true ->
+            throw({error, different_table_schema});
+        false ->
+            ok
+    end.
+
+convert_copies({K, [_ | _]}) when K == ram_copies; K == disc_copies; K == disc_only_copies ->
+    {K, [node()]};
+convert_copies(Other) ->
+    Other.
+
+extract_backup(BackupFileName) ->
+    BackupDir = root_backup_dir(),
+    ok = validate_filenames(BackupFileName),
+    ?fmt_tar_err(erl_tar:extract(BackupFileName, [{cwd, BackupDir}, compressed])).
+
+validate_filenames(BackupFileName) ->
+    {ok, FileNames} = ?fmt_tar_err(erl_tar:table(BackupFileName, [compressed])),
+    BackupName = filename:basename(BackupFileName, ?TAR_SUFFIX),
+    IsValid = lists:all(
+        fun(FileName) ->
+            [Root | _] = filename:split(FileName),
+            Root =:= BackupName
+        end,
+        FileNames
+    ),
+    case IsValid of
+        true -> ok;
+        false -> {error, bad_archive_dir}
+    end.
+
+import_cluster_hocon(BackupDir, Opts) ->
+    HoconFileName = filename:join(BackupDir, ?CLUSTER_HOCON_FILENAME),
+    case filelib:is_regular(HoconFileName) of
+        true ->
+            {ok, RawConf} = hocon:files([HoconFileName]),
+            {ok, _} = validate_cluster_hocon(RawConf),
+            maybe_print("Importing cluster configuration...~n", [], Opts),
+            %% At this point, when all validations have been passed, we want to log errors (if any)
+            %% but proceed with the next items, instead of aborting the whole import operation
+            do_import_conf(RawConf, Opts);
+        false ->
+            maybe_print("No cluster configuration to be imported.~n", [], Opts),
+            ?SLOG(info, #{
+                msg => "no_backup_hocon_config_to_import",
+                backup => BackupDir
+            }),
+            #{}
+    end.
+
+read_data_files(RawConf) ->
+    DataDir = bin(emqx:data_dir()),
+    RawConf1 = emqx_authz:maybe_read_acl_file(RawConf),
+    emqx_utils_maps:deep_convert(RawConf1, fun read_data_file/3, [DataDir]).
+
+read_data_file(Key, Val, DataDir) ->
+    Val1 =
+        case Val of
+            <<DataDir:(byte_size(DataDir))/binary, _RestPath/binary>> = FileName ->
+                case file:read_file(FileName) of
+                    {ok, Content} ->
+                        Content;
+                    {error, Reason} ->
+                        ?SLOG(warning, #{
+                            msg => "failed_to_read_data_file",
+                            filename => FileName,
+                            reason => Reason
+                        }),
+                        FileName
+                end;
+            V ->
+                V
+        end,
+    {Key, Val1}.
+
+validate_cluster_hocon(RawConf) ->
+    %% write ACL file to comply with the schema...
+    RawConf1 = emqx_authz:maybe_write_acl_file(RawConf),
+    emqx_hocon:check(
+        emqx_conf:schema_module(),
+        maps:merge(emqx:get_raw_config([]), RawConf1),
```

Comment:
The idea is to validate the whole config before trying to import its individual root keys.
I use `maps:merge/2` just to put existing values for required keys like `node` (no need for deep merging). 
I wonder if there are better alternatives.. :thinking:

Reply:
Discussed and agreed that it must be fine.